uring/uring_test.go

```diff
@@ -0,0 +1,269 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package uring
+
+import (
+	"errors"
+	"io/ioutil"
+	"math"
+	"os"
+	"runtime"
+	"syscall"
+	"testing"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+const openFile = "./../go.mod"
+
+func MustNil(t *testing.T, val interface{}) {
+	t.Helper()
+	Assert(t, val == nil, val)
+	if val != nil {
+		t.Fatal("assertion nil failed, val=", val)
+	}
+}
+
+func MustTrue(t *testing.T, cond bool) {
+	t.Helper()
+	if !cond {
+		t.Fatal("assertion true failed.")
+	}
+}
+
+func Equal(t *testing.T, got, expect interface{}) {
+	t.Helper()
+	if got != expect {
+		t.Fatalf("assertion equal failed, got=[%v], expect=[%v]", got, expect)
+	}
+}
+
+func Assert(t *testing.T, cond bool, val ...interface{}) {
+	t.Helper()
+	if !cond {
+		if len(val) > 0 {
+			val = append([]interface{}{"assertion failed:"}, val...)
+			t.Fatal(val...)
+		} else {
+			t.Fatal("assertion failed")
+		}
+	}
+}
+
+func TestClose(t *testing.T) {
+	u, err := IOURing(8)
+	MustNil(t, err)
+	Assert(t, u.Fd() != 0)
+	defer u.Close()
+
+	f, err := os.Open(openFile)
+	MustNil(t, err)
+	defer f.Close()
+
+	err = u.Queue(Close(f.Fd()), 0, 0)
+	MustNil(t, err)
+
+	_, err = u.Submit()
+	MustNil(t, err)
+
+	cqe, err := u.WaitCQE()
+	MustNil(t, err)
+	MustNil(t, cqe.Error())
+
+	_, err = unix.FcntlInt(f.Fd(), unix.F_GETFD, 0)
+	Equal(t, err, unix.EBADF)
+}
+
+func TestReadV(t *testing.T) {
+	u, err := IOURing(8)
+	MustNil(t, err)
+	defer u.Close()
+
+	f, err := os.Open(openFile)
+	MustNil(t, err)
+	defer f.Close()
+
+	v, err := makeV(f, 16)
+	MustNil(t, err)
+
+	err = u.Queue(ReadV(f.Fd(), v, 0), 0, 0)
+	MustNil(t, err)
+
+	_, err = u.Submit()
+	MustNil(t, err)
+
+	cqe, err := u.WaitCQE()
+	MustNil(t, err)
+	MustNil(t, cqe.Error())
+
+	expected, err := ioutil.ReadFile(openFile)
+	MustNil(t, err)
+	Assert(t, vToString(v) == string(expected))
+}
+
+func TestReady(t *testing.T) {
+	u, err := IOURing(8)
+	MustNil(t, err)
+	defer u.Close()
+
+	Equal(t, u.cqRing.ready(), uint32(0))
+
+	err = queueSQEs(u, 5, 0)
+	Equal(t, u.cqRing.ready(), uint32(5))
+
+	u.CQESeen()
+	Equal(t, u.cqRing.ready(), uint32(4))
+
+	u.Advance(4)
+	Equal(t, u.cqRing.ready(), uint32(0))
+}
+
+func TestTimeoutWait(t *testing.T) {
+	u, err := IOURing(8)
+	MustNil(t, err)
+	defer u.Close()
+
+	err = u.Queue(Nop(), 0, 1)
+	MustNil(t, err)
+
+	if u.Params.features&IORING_FEAT_EXT_ARG != 0 {
+		n, err := u.Submit()
+		MustNil(t, err)
+		Equal(t, n, uint(1))
+	}
+
+	n := 0
+	for {
+		cqe, err := u.WaitCQETimeout(time.Second)
+		if errors.Is(err, syscall.ETIME) {
+			break
+		}
+		if errors.Is(err, syscall.EINTR) || errors.Is(err, syscall.EAGAIN) {
+			runtime.Gosched()
+			continue
+		}
+
+		MustNil(t, err)
+		u.CQESeen()
+
+		MustNil(t, cqe.Error())
```

Comment:
2nd iteration got timeout assertion

Reply:
Thx for your reminder, I'm stuck here and looking for a solution. ;)
```suggestion
		// TODO: Add println to make timer expired
		println("SysEnter in")
		ret, err := SysEnter(u.fd, data.submit, data.waitNr, flags, data.arg, data.sz)
		println("SysEnter out")
```

Reply:
This is because on Linux kernel version above 5.10 is using new io_uring timeout (feat `IORING_FEAT_EXT_ARG` is on set), it expected to throw timeout error on https://github.com/cloudwego/netpoll/pull/197/files/381627225b3c51380f7db5869b0d3d63ac0fd3b3#diff-bf1a7df24d2545c5374112650bc1f0aeef7e1065fa2abd4c2becdce9e3bb7adcR142-R143.

So this test (current `TestTimeoutWait`) would pass on older kernel version I suppose.

Reply:
You've got a typo,
This one should be checking the ring feature, and according to liburing, it only enqueue timeout entries if timeout is set.
https://github.com/cloudwego/netpoll/pull/197/files/381627225b3c51380f7db5869b0d3d63ac0fd3b3#diff-e670f6a8a5026481de5f93cf93d3e5b7ed7f043102f37d690c2a34c83ce215ecR131

Reply:
Thx a lot, it was an oversight on my part, but it still seems not to pass. I'll fix this in my local first. )